[decisions] keyset types

kodebach opened the pull request:
<!--
Check relevant points but **please do not remove entries**.
-->

## Basics

<!--
These points need to be fulfilled for every PR.
-->

- [ ] Short descriptions of your changes are in the release notes
      (added as entry in `doc/news/_preparation_next_release.md` which
      contains `_(my name)_`)
      **Please always add something to the release notes.**
- [ ] Details of what you changed are in commit messages
      (first line should have `module: short statement` syntax)
- [ ] References to issues, e.g. `close #X`, are in the commit messages.
- [ ] The buildservers are happy. If not, fix **in this order**:
  - [ ] add a line in `doc/news/_preparation_next_release.md`
  - [ ] reformat the code with `scripts/dev/reformat-all`
  - [ ] make all unit tests pass
  - [ ] fix all memleaks
- [ ] The PR is rebased with current master.

<!--
If you have any troubles fulfilling these criteria, please write
about the trouble as comment in the PR. We will help you,
but we cannot accept PRs that do not fulfill the basics.
-->

## Checklist

<!--
For docu fixes, spell checking, and similar none of these points below
need to be checked.
-->

- [ ] I added unit tests for my code
- [ ] I fully described what my PR does in the documentation
      (not in the PR description)
- [ ] I fixed all affected documentation (see [Documentation Guidelines](https://www.libelektra.org/devgettingstarted/documentation))
- [ ] I fixed all affected decisions (see [Decision Process](https://www.libelektra.org/decisions/decision-process))
- [ ] I added code comments, logging, and assertions as appropriate (see [Coding Guidelines](https://www.libelektra.org/devgettingstarted/coding))
- [ ] I updated all meta data (e.g. README.md of plugins and [METADATA.ini](https://master.libelektra.org/doc/METADATA.ini))
- [ ] I mentioned [every code](/.reuse/dep5) not directly written by me in [reuse syntax](https://reuse.software/)

## Review

<!--
Reviewers should check the following.
-->

- [ ] Documentation is introductory, concise, good to read and describes everything what the PR does
- [ ] Examples are well chosen and understandable
- [ ] Code is conforming to [our Coding Guidelines](https://master.libelektra.org/doc/CODING.md)
- [ ] APIs are conforming to [our Design Guidelines](https://master.libelektra.org/doc/DESIGN.md)
- [ ] Code is consistent to [our Design Decisions](https://master.libelektra.org/doc/decisions)

## Labels

<!--
If you are already Elektra developer, please adjust the labels.
Otherwise, write a comment and it will be done for you.
-->

- [ ] Add the "work in progress" label if you do not want the PR to be reviewed yet.
- [ ] Add the "ready to merge" label **if the basics are fulfilled** and no further pushes are planned by you.

kodebach commented:
Not complete yet. I need to think about the metadata stuff in relation to COW. Both decisions are in one PR, because the metadata stuff definitely depends on the keyset types.

kodebach commented:
I decided to move `metakey_semantics` to a separate PR after all. It is not as closely related as I thought.

This decision here is very straightforward IMO, so I hope we can approve it very quickly. Maybe we can even directly go to `3_decided` in this PR.

markus2330 commented on doc/decisions/0_drafts/keyset_types.md:
```diff
@@ -0,0 +1,79 @@
+A `KeySet` can be used in different ways.
+Among these are at least:
+
+- As originally intended, a set of configuration data
+- Metadata of a `Key`
+- General associative array data structure
```
This would be perfect to have in use cases.

kodebach replied:
https://github.com/ElektraInitiative/libelektra/blob/3e9036fc91533cc03dfbb6c3fc0a01cf45f12788/doc/usecases/core/README.md#L5-L12

markus2330 commented on doc/decisions/0_drafts/keyset_types.md:
```diff
@@ -0,0 +1,79 @@
+
+## Constraints
+
+- API duplication just to separate types should be avoided
```
I agree about the goal but I don't think its a constraint.

```suggestion
- API must be minimal
```

kodebach replied:
You could argue the API would still be minimal, since you cannot reduce it without loosing type safety or functionality.

kodebach replied:
Maybe we make it an assumption:

"Introducing a type safe API with different C-types for each type of `KeySet`, is not worth the necessary API duplication"

markus2330 commented on doc/decisions/0_drafts/keyset_types.md:
```diff
@@ -0,0 +1,79 @@
+
+## Assumptions
+
+- The types are not different enough to warrant the overhead of entirely different APIs and data structures.
```
Assumption is okay :+1:

markus2330 commented on doc/decisions/0_drafts/keyset_types.md:
```diff
@@ -0,0 +1,79 @@
+
+Leave `KeySet` entirely unrestricted, so it can be used for each use case.
+
+Sometimes there can be odd edge cases.
+For example, what should a storage plugin do, if it receives a `KeySet` without both configuration data and metadata.
```
Put this up to the problem.

kodebach replied:
Just noticed a typo here:

```suggestion
Sometimes there can be odd edge cases.
For example, what should a storage plugin do, if it receives a `KeySet` with both configuration data and metadata.
```

I'll add it to the problem as well.

markus2330 commented on doc/decisions/0_drafts/keyset_types.md:
```diff
@@ -0,0 +1,79 @@
+1. Metadata:
+   Can only contain `Key`s with namespace `KEY_NS_META`.
+2. General data:
+   Can only contain `Key`s with namespace `KEY_NS_DATA`.
```
KEY_NS_DATA should be explained somewhere, probably best in its use case.

kodebach replied:
I think I'll add a new decision to this PR that introduces `KEY_NS_DATA`. It should be a very short and simple decision and under the constraint of this decision there is only one alternative: "Don't use `KeySet` for general data".

markus2330 commented on doc/decisions/0_drafts/keyset_types.md:
```diff
@@ -0,0 +1,79 @@
+
+To effectively create different kinds of `KeySet`s, we restrict which `Key`s can be inserted into a `KeySet` based on the first `Key` in a `KeySet`.
+
+Any `Key` can be inserted into an empty `KeySet`.
```
```suggestion
Any `Key` except cascading keys can be inserted into an empty `KeySet`.
```

markus2330 commented on doc/decisions/0_drafts/keyset_types.md:
```diff
@@ -0,0 +1,79 @@
+
+Importantly, `Key`s with namespace `KEY_NS_CASCADING` can never be inserted into a `KeySet`.
+This is a restriction that simplifies the logic for cascading lookups.
+This can now be achieved with the new `KEY_NS_DATA` namespace.
```
Please clarify what "This" is.

kodebach replied:
```suggestion
Instead the `KEY_NS_DATA` namespace should be used.
```

markus2330 commented on doc/decisions/0_drafts/keyset_types.md:
```diff
@@ -0,0 +1,79 @@
+Expose a different type and API for each of the use cases.
+
+This creates unnecessary overhead and at least some code duplication, even if it is minimized by extracting the implementation.
+
```
```suggestion

### KeySets also have Namespaces

Add a `ksSetNamespace` API which allows to change the namespace of a KeySet.
This requires an alias `KEY_NS_CONFIG` to the namespaces `KEY_NS_SPEC`, `KEY_NS_PROC`, `KEY_NS_DIR`, `KEY_NS_USER`, `KEY_NS_SYSTEM` and `KEY_NS_DEFAULT`.

Only keys of the namespace set to `ksSetNamespace` can be added.
`ksSetNamespace` fails if called after keys were added.

This alternative conflicts with the constraint of minimal API.
```

kodebach replied:
It shouldn't be `KEY_NS_CONFIG`. There should be a separate enum

```c
enum KeySetType {
    KS_TYPE_NONE,
    KS_TYPE_CONFIG,
    KS_TYPE_META,
    KS_TYPE_DATA
};
```

markus2330 commented on doc/decisions/0_drafts/keyset_types.md:
```diff
@@ -0,0 +1,79 @@
+
+## Rationale
+
+Some kind of separation is necessary to avoid issues with e.g. `KeySet * keyMeta(Key * key)`
+This way we avoid API/code duplication and still allow slightly different handling of the different types.
```
Better avoid Rationale now, as it is unclear to what it refers.

Please describe a solution for metadata above in the "Considered Alternatives". We obviously don't want that someone removes all keys and mutates a meta-data keyset to some other keyset later on.

kodebach replied:
> We obviously don't want that someone removes all keys and mutates a meta-data keyset to some other keyset later on.

Hm, yeah that's an issue. How about we add a flag

```
bool forceMeta : 1; /**< if set the KeySet can only contain meta:/ keys */
```

markus2330 reviewed (approved):
Great write-up! I added a few suggestions, pls fix then we can merge it as "in progress" (or "in review" with new names for steps).

kodebach commented:
@markus2330 I updated the `keyset_types` decision with the review suggestions and moved it to `2_in_progress`. I also added a new `data_namespace` decision in directly in `3_decided`.

markus2330 reviewed (approved):
- There is a wrong link, please make sure this doesn't happen before requesting reviews as it is one more round-trip.
- Please don't intermix different target steps within one decision in future. Especially that a later added decision gets a higher target is irritating.
- The decisions themselves, however, are imho okay. Lets look for a second reviewer. Maybe someone finds a better name as KEY_NS_DATA?

kodebach commented:
> Maybe someone finds a better name as `KEY_NS_DATA`?

What could be a better name for something that is intended for general data?

kodebach commented:
jenkins build libelektra please

atmaxinger reviewed (approved):
All in all, I think this decision makes sense. 

I also thought about it from a recording standpoint, and I don't think it would be impacted much.

As to what is the best solution - I'm not sure yet. Personally, I think 'Keysets also have Namespaces' would be the best, as "type safety" can be guaranteed best with it. The implicit 'Restriction based on first Key' is a close second.

As for `KEY_NS_DATA`, maybe we could use `KEY_NS_MISC` for miscellaneous data?

kodebach commented:
I like the `KEY_NS_MISC` idea. I'll change it.

kodebach commented:
@markus2330 I believe this can be merged as soon as the CI succeeds.

The goals of these decisions are:
- `misc_namespace`: Since it's already in decided (and approved), I don't think this requires any extra explanation.
- `keyset_types`: As a results of `misc_namespace` and because of `keyMeta()` we should separate different kinds of `KeySet`. This is a new decision, to determine how to best do that.

atmaxinger commented on doc/decisions/0_drafts/metakey_semantics.md:
```diff
@@ -252,5 +252,6 @@ Therefore, the `keyData` pointers will only be the same if the value was not mod
 ## Related Decisions
 
 - [Copy On Write](../2_in_progress/copy_on_write.md)
+- [Types of `KeySet`s](../0_drafts/keyset_types.md)
```
```suggestion
- [Types of `KeySet`s](../2_in_progress/keyset_types.md)
```

atmaxinger commented on doc/decisions/3_decided/misc_namespace.md:
```diff
@@ -0,0 +1,65 @@
+Introduce separate `KEY_NS_MISC` namespace.
+
+In the unescaped form this will be `\x09` (byte with decimal value `9`).
+In the escaped form it will be `data:/`.
```
Is there a reason why the escaped form is not `misc:/`?

kodebach replied:
Good catch, seems I forgot to change it

kodebach replied:
```suggestion
In the escaped form it will be `misc:/`.
```

kodebach commented:
@markus2330 I think this is ready to merge now.
